Add a slew of validations.

jgkamat opened the pull request:
Closes #195

Adds a bunch of validations for a couple functions. I think I've got the desired behavior except for `(duration #)`. For some reason, the assertion only works on `(set-duration #)`, and everything I tried to rename it to duration (and alias to set-duration) dosen't seem to work. I'm hoping you'll be able to figure out what I'm doing wrong there. Seems like things like `a-1` still work, but `a0` is caught. 

Let me know if I'm doing anything wrong, this is my first time writing clojure :smiley:

daveyarwood commented:
This looks great! I only see a few issues, and they are mostly just style points. I will leave comments on individual lines.

The duration issue is a result of this hack that we can remove now: https://github.com/alda-lang/alda/blob/master/server/src/alda/lisp/model/duration.clj#L73  `duration` calls `set-duration` under the hood, but only if the duration is positive. Since we are doing all the validation in `set-duration` now, we can change this line to just `(set-duration beats)`.

daveyarwood commented on server/src/alda/lisp/attributes.clj:
```diff
@@ -14,15 +14,28 @@
   (constantly (/ x 100.0)))
 
+;; Validation that the input is an integer value
+(defn- pos-integer[x]
```
There should be a space between the function name and the vector of arguments: `pos-integer [x]`

daveyarwood commented on server/src/alda/lisp/attributes.clj:
```diff
@@ -14,15 +14,28 @@
+                        (map? val)
+                        (and (number? val)
+                             (pos? val)))]}
+
```
1. In a Clojure map, there should usually be no space between the opening curly brace and the first thing in it.

2. Re: line breaks / spacing, this is a matter of preference, but I would write it like this:

```clojure
(fn [val]
  {:pre [(or (map? val)
             (and (number? val) (pos? val)))]}
```

daveyarwood commented on server/src/alda/lisp/attributes.clj:
```diff
@@ -69,6 +82,22 @@
+  (let [clean-str (apply str (filter #(Character/isLetter %) key-sig))]
+    (and
+     ;; Check to see if any chars are not a-g
+     (empty? (re-find #"[^a-gA-G]" clean-str))
```
1. We should actually only allow `a-g`. Something like `"F+ C+ G+"` should throw an error.

2. `re-find` will return `nil` if it doesn't find any instances of the regex in the string. So, it would be more idiomatic to use `not` here, instead of `empty?` which has slightly different semantics.

3. This is also a matter of personal preference, but I would include the first condition on the same line of the `and`, like this:

    ```clojure
    (and (not (re-find #"[^a-g]" clean-str))
         (= (count (distinct clean-str)) (count clean-str)))
    ```

elyisgreat commented:
Can someone summarize the new validations? I don't speak clojure :(

jgkamat commented:
Here's whats changed (still working on the finishing touches):
Errors on:
```
(key-sig "a+ a+")
(key-sig "z+")
(duration -1)
(duration 0)
(octave 1.1)
(tempo 0)
(tempo -1)
(tempo 1.1)
```
While keeping the rest of the behavior intact

elyisgreat commented:
Tempo 1.1 is not a problem.

daveyarwood commented:
@elyisgreat You're right, arguably we should allow non-integers for tempo. The key thing is just that tempo is a positive number.

jgkamat commented:
I'll fix that in my next commit, thanks!

jgkamat commented:
I think I addressed all the problems, but the check for `(duration 0)` seemed to be causing problems (so I removed it), when I ran `(duration (ms 100))` that check would cause it to fail. If you can tell me what I'm doing wrong, I'll fix it! :smile: 

The tests seem to be passing for me now.

daveyarwood commented on server/src/alda/lisp/model/duration.clj:
```diff
@@ -70,7 +70,7 @@
                              :when (= type :milliseconds)]
                          value))]
-    (when (pos? beats) (set-duration beats))
+    (set-duration beats)
```
I didn't think about millisecond durations. This is something we should probably address in the future, but for now, we can continue to ignore them.

I think changing this line to `(when beats (set-duration beats))` will do the trick.

jgkamat commented:
Sorry for the spotty turnarounds, I've been kind of busy today. I haven't gotten a chance to test the latest commit at all (I can't seem to get the build working on my second laptop, i'll debug that later on). I'll test this out when I get back, and comment what I find out.

jgkamat commented:
The `(duration (ms 100))` test seems to be failing, but otherwise, it looks good.

daveyarwood commented:
@jgkamat I just played around with it, and I've got a fix ready. I'm going to merge your PR, and then apply my change in a new commit.

The issue was that beats is never actually `nil` -- it defaults to 0 if there are no note-lengths expressed in beats (for example if you do `(duration (ms 1000))`). So if we do `(when beats (set-duration beats))`, then it will try to set the duration to 0 beats, triggering the assertion error because the number isn't positive.

Thanks for this PR, nice work! :+1: :smile: :+1:
